pkg/cache/resource_node.go

```diff
@@ -0,0 +1,157 @@
+	getResourceNode() ResourceNode
+
+	HasParent() bool
+	Parent() T
+}
+
+// available determines how much capacity remains for the current
+// node, taking into account usage and BorrowingLimits.
+//
+// This function may return a negative number in the case of
+// overadmission - e.g. capacity was removed or the node moved to
+// another Cohort.
+//
+// We add an option to ignore the borrowing limit, to support
+// features.MultiplePreemptions=false path. This will be cleaned up in
+// v0.10, when we delete legacy logic.
+func available[T wiredResourceNode[T]](node wiredResourceNode[T], fr resources.FlavorResource, enforceBorrowLimit bool) int64 {
+	r := node.getResourceNode()
+	if !node.HasParent() {
+		return r.SubtreeQuota[fr] - r.Usage[fr]
+	}
+	localAvailable := max(0, r.guaranteedQuota(fr)-r.Usage[fr])
+	parentAvailable := available(node.Parent(), fr, enforceBorrowLimit)
+
+	if borrowingLimit := r.Quotas[fr].borrowingLimitSafe(); enforceBorrowLimit && borrowingLimit != nil {
+		storedInParent := r.SubtreeQuota[fr] - r.guaranteedQuota(fr)
+		usedInParent := max(0, r.Usage[fr]-r.guaranteedQuota(fr))
+		withMaxFromParent := storedInParent - usedInParent + *borrowingLimit
+		parentAvailable = min(withMaxFromParent, parentAvailable)
+	}
+	return localAvailable + parentAvailable
+}
+
+// maxAvailable returns the maximum capacity available to this node,
+// assuming no usage, while respecting BorrowingLimits.
+func maxAvailable[T wiredResourceNode[T]](node wiredResourceNode[T], fr resources.FlavorResource) int64 {
+	r := node.getResourceNode()
+	if !node.HasParent() {
+		return r.SubtreeQuota[fr]
+	}
+	available := r.guaranteedQuota(fr) + maxAvailable(node.Parent(), fr)
+	if borrowingLimit := r.Quotas[fr].borrowingLimitSafe(); borrowingLimit != nil {
+		maxWithBorrowing := r.SubtreeQuota[fr] + *borrowingLimit
+		available = min(maxWithBorrowing, available)
+	}
+	return available
+}
+
+// addUsage adds usage to the current node, and bubbles up usage to
+// its Cohort when usage exceeds guaranteedQuota.
+func addUsage[T wiredResourceNode[T]](node wiredResourceNode[T], fr resources.FlavorResource, val int64) {
+	r := node.getResourceNode()
+	localAvailable := max(0, r.guaranteedQuota(fr)-r.Usage[fr])
+	r.Usage[fr] += val
+	if node.HasParent() && val > localAvailable {
+		deltaParentUsage := val - localAvailable
+		addUsage(node.Parent(), fr, deltaParentUsage)
+	}
+}
+
+// removeUsage removes usage from the current node, and removes usage
+// past guaranteedQuota that it was storing in its Cohort.
+func removeUsage[T wiredResourceNode[T]](node wiredResourceNode[T], fr resources.FlavorResource, val int64) {
+	r := node.getResourceNode()
+	usageStoredInParent := r.Usage[fr] - r.guaranteedQuota(fr)
+	r.Usage[fr] -= val
+	if usageStoredInParent <= 0 || !node.HasParent() {
+		return
+	}
+	deltaParentUsage := min(val, usageStoredInParent)
+	removeUsage(node.Parent(), fr, deltaParentUsage)
+}
+
+func updateClusterQueueResourceNode(cq *clusterQueue) {
```

Comment:
it looks like this should live in clusterqueue.go, and probably a member function.